pkg/http/listener.go

```diff
@@ -0,0 +1,294 @@
+		// Just drop the connection.
+		if listener.tlsConfig == nil {
+			if listener.errorLogFunc != nil {
+				listener.errorLogFunc(errors.New("junk message"),
+					"Received non-HTTP message from new connection %s at server %s",
+					bufconn.RemoteAddr(), bufconn.LocalAddr())
+			}
+			bufconn.Close()
+			return
+		}
+
+		// As the listener is configured with TLS, try to do TLS handshake, drop the connection if it fails.
+		tlsConn := tls.Server(bufconn, listener.tlsConfig)
+		if err = tlsConn.Handshake(); err != nil {
+			if listener.errorLogFunc != nil {
+				listener.errorLogFunc(err,
+					"TLS handshake failed with new connection %s at server %s",
+					bufconn.RemoteAddr(), bufconn.LocalAddr())
+			}
+			tlsConn.Close()
+			return
+		}
+
+		// As TLS handshake is successful, return TLSConn.
+		send(acceptResult{tlsConn, nil}, doneCh)
+		return
+	}
+
+	// Closure to handle new connections till done channel is not closed.
+	handleConnection := func(tcpListener *net.TCPListener, doneCh <-chan struct{}) {
+		for {
+			tcpConn, err := tcpListener.AcceptTCP()
+			if err != nil {
+				// Returns when send fails.
+				if !send(acceptResult{nil, err}, doneCh) {
+					return
+				}
+			} else {
+				go acceptTCP(tcpConn, doneCh)
+			}
+		}
+	}
+
+	// Start separate goroutine for each TCP listener to handle connection.
+	for _, tcpListener := range listener.tcpListeners {
+		doneCh := make(chan struct{})
+		go handleConnection(tcpListener, doneCh)
+		listener.doneChs = append(listener.doneChs, doneCh)
+	}
+}
+
+// Accept - reads from httpListener.acceptCh for one of previously accepted TCP connection and returns the same.
+func (listener *httpListener) Accept() (conn net.Conn, err error) {
+	result, ok := <-listener.acceptCh
+	if ok {
+		return result.conn, result.err
+	}
+
+	return nil, io.EOF
+}
+
+// Close - closes underneath all TCP listeners.
+func (listener *httpListener) Close() (err error) {
+	listener.mutex.Lock()
+	defer listener.mutex.Unlock()
+	if listener.doneChs == nil {
+		return nil
+	}
+
+	wg := &sync.WaitGroup{}
+	for i := range listener.tcpListeners {
+		wg.Add(1)
+		go func(tcpListener *net.TCPListener, doneCh chan<- struct{}) {
+			tcpListener.Close()
+			close(doneCh)
+			wg.Done()
+		}(listener.tcpListeners[i], listener.doneChs[i])
+	}
+	wg.Wait()
+
+	listener.doneChs = nil
+	return nil
+}
+
+// Addr - net.Listener interface compatible method returns net.Addr.  In case of multiple TCP listeners, it returns '0.0.0.0' as IP address.
+func (listener *httpListener) Addr() (addr net.Addr) {
+	addr = listener.tcpListeners[0].Addr()
+	if len(listener.tcpListeners) == 1 {
+		return addr
+	}
+
+	tcpAddr := addr.(*net.TCPAddr)
+	if ip := net.ParseIP("0.0.0.0"); ip != nil {
+		tcpAddr.IP = ip
+	}
+
+	addr = tcpAddr
+	return addr
+}
+
+// Addrs - returns all address information of TCP listeners.
+func (listener *httpListener) Addrs() (addrs []net.Addr) {
```

Comment:
`net.Listener` interface returns only single address with `Addr()` function whereas this `httpListener` handles multiple addresses, `Addrs()` function is enhanced version returns multiple net.Addr in an array.